File: src/core/progress.ts
/** What to show for one step of the refresh. */
export interface ProgressStep {
  message: string;
  /** Percentage points to add to the bar since the previous step. */
  increment: number;
}

/**
 * Turns "checked 3 of 40" into the increments a progress bar wants.
 *
 * The bar takes deltas, not totals, so it only fills correctly if each step
 * reports the difference from the last one. Increments across a whole run add
 * up to exactly 100.
 */
export class ProgressTracker {
  private reportedPercent = 0;

  step(checked: number, total: number): ProgressStep {
    const percent = total > 0 ? (checked / total) * 100 : 100;
    const increment = percent - this.reportedPercent;
    this.reportedPercent = percent;

    return {
      message: `${checked} of ${total} packages checked`,
      increment
    };
  }
}

File: clients/intellij/src/main/kotlin/dev/pubgrade/core/Progress.kt
package dev.pubgrade.core

/** What to show for one step of the refresh. */
data class ProgressStep(
    val message: String,
    /** How full the bar should be, 0.0 to 1.0. */
    val fraction: Double
)

/**
 * Turns "checked 3 of 40" into what a progress bar wants.
 *
 * The VS Code client reports deltas here because its API takes increments.
 * IntelliJ's indicator takes an absolute fraction instead, so this is the one
 * piece of the port that is genuinely different rather than translated.
 */
fun progressStep(checked: Int, total: Int) = ProgressStep(
    message = "$checked of $total packages checked",
    fraction = if (total > 0) checked.toDouble() / total else 1.0
)
